forc-plugins/forc-fmt/src/main.rs

```diff
@@ -69,29 +77,177 @@ fn run() -> Result<()> {
     };
 
     let manifest_file = forc_pkg::manifest::ManifestFile::from_dir(&dir)?;
-    for (_, member_path) in manifest_file.member_manifests()? {
-        let member_dir = member_path.dir();
-        let mut formatter = Formatter::from_dir(member_dir)?;
-        format_pkg_at_dir(&app, &dir, &mut formatter)?;
+
+    match manifest_file {
+        ManifestFile::Workspace(ws) => {
+            format_workspace_at_dir(&app, &ws, &dir)?;
+        }
+        ManifestFile::Package(_) => {
+            let mut formatter = Formatter::from_dir(&dir)?;
+            format_pkg_at_dir(&app, &dir, &mut formatter)?;
+        }
     }
 
     Ok(())
 }
 
-fn format_single_file(path: &Path) -> Result<()> {
-    let mut formatter = Formatter::default();
-    if let Ok(file_content) = fs::read_to_string(path) {
-        match Formatter::format(&mut formatter, file_content.into(), None) {
+/// Recursively get a Vec<PathBuf> of subdirectories that contains a Forc.toml.
+fn get_sway_dirs(workspace_dir: PathBuf) -> Vec<PathBuf> {
+    let mut dirs_to_format = vec![];
+    let mut dirs_to_search = vec![workspace_dir];
+
+    while let Some(next_dir) = dirs_to_search.pop() {
+        if let Ok(read_dir) = fs::read_dir(next_dir) {
+            for entry in read_dir.filter_map(|res| res.ok()) {
+                let path = entry.path();
+
+                if path.is_dir() {
+                    dirs_to_search.push(path.clone());
+                    if path.join(constants::MANIFEST_FILE_NAME).exists() {
+                        dirs_to_format.push(path);
+                    }
```

Comment:
Is this a valid directory structure? Say I have a project called `fizzbuzz` with the following structure:

```
.
├── Forc.lock
├── Forc.toml
├── proj1
│   ├── Forc.toml
│   └── src
│       └── main.sw
└── src
    └── main.sw
```

If so, I think this will still format sub projects twice for nested sway projects. In this example,

`get_sway_dirs` will return: `[fizzbuzz, proj1]`

`proj1` will get formatted twice, once with the config from `proj1`'s manifest and once with the config from `fizzbuzz`'s manifest.

One way to prevent this would be, in `format_pkg_at_dir`, to keep track of which files have been formatted and skip them if so. I think you'd also need to iterate through the result of `get_sway_dirs` in reverse to ensure the nested subdirectory manifest was used for formatting before the outer directory's manifest.

```
for sub_dir in get_sway_dirs(dir.to_path_buf()).reverse()
```

Same thing for `format_workspace_at_dir`

Reply:
Good catch! If there are nested packages within a package when handling `format_pkg_at_dir()` this would certainly be an issue, though I'm not sure if it's valid to do that. `forc` allows me to build a random project i spun up, so that means it must be allowed, and we should probably handle this scenario. I'll go with your suggestion to fix that!

However, in the case of workspace formatting, since we're starting at with `workspace_dir` already inside `dirs_to_search`, `workspace_dir` itself isn't added to `dirs_to_format`! ;) Of course, this assumes that there are no .sw files in the root, hence the code block [here](https://github.com/FuelLabs/sway/pull/4038/files#diff-bf9b5d7023fff817f83ae31bd01e66d2788178a0705a2c638ce79d7990cc374aR174-R188) that handles that assumption. So I think the possible multiple formats can occur at the package level and not the workspace level, which `get_sway_dirs()` doesn't handle.

Reply:
Opened https://github.com/FuelLabs/sway/issues/4050 to track this